Support for GitHub Models using the Azure AI Inference API

jdubois opened the pull request:
Fix #1719

This adds GitHub Models (see https://github.com/marketplace/models ) support with the new Azure AI Inference API Java SDK.

Documentation: https://docs.langchain4j.dev/integrations/language-models/github-models

jdubois committed:
First working test using the unreleased Azure AI Inference API

Fix #1719

jdubois committed:
Merge branch 'main' into fix-1719

maxandersen commented on langchain4j-github-models/pom.xml:
```diff
@@ -0,0 +1,120 @@
+
+        <dependency>
+            <groupId>com.azure</groupId>
+            <artifactId>azure-ai-inference</artifactId>
```
I thought GitHub models exposed openai apis - can't we reuse existing api ? Or is there a value add of adding  such dependency?

jdubois replied:
Have a look at https://github.com/marketplace/models/azure-openai/gpt-4o in "get started", and check the API in JavaScript (Java will be added soon): you have the choice between OpenAI and the Azure AI Inference API.
So yes, the OpenAI models should work for accessing OpenAI models.
Then, for https://github.com/marketplace/models/azureml-mistral/Mistral-Nemo you have the choice between the Mistral AI SDK and the Azure AI Inference API.
And for https://github.com/marketplace/models/azureml/Phi-3-5-mini-instruct it will be only the Azure AI Inference API.
So the Azure AI Inference API is the only one which should work across all models. I agree, differences are probably very light, but that might change.

n3kh3l replied:
In "Get Started," if we check the API specifications for REST—whether it's "gpt-4o," "Mistral-Nemo," or "Phi-3.5-mini-instruct"—the API specifications are the same for all and are compatible with OpenAI's API, which uses a Bearer token. Therefore, if you use langchain4j's OpenAiChatModel API, it will work with all the supported GitHub models. The same code snippet I mentioned in issue(https://github.com/langchain4j/langchain4j/issues/1719#issuecomment-2361872630)  works for all the GitHub models you referenced.

Now, regarding the "Azure AI Inference API," you’re correct that it may not currently support Java. However, langchain4j's OpenAiChatModel API alone suffices for your requirements to use GitHub models. Internally, it does not rely on any SDKs like the Mistral AI SDK or the Azure AI Inference SDK.

Aside from this, is there any specific requirement that langchain4j's OpenAiChatModel API cannot address?

jdubois replied:
Oh yes: it’s not done by the companies doing those models, so you can’t rely on them for maintenance, support, new features… Most companies will refuse to use the non official SDK, and I can confirm this on the Azure OpenAI side where I have the LangChain4J OpenAI and the LangChain4J Azure OpenAI stats, which is exactly the same situation - nearly everyone uses the official SDK.
The Java version of the Azure AI Inference API had its first release yesterday.

jdubois replied:
I will also have access to the GitHub Models stats, so I’ll be able to tell which implementation people use.
If this implementation isn’t popular, you can trust me that I will remove this module.

n3kh3l replied:
> Oh yes: it’s not done by the companies doing those models, so you can’t rely on them for maintenance, support, new features… Most companies will refuse to use the non official SDK, and I can confirm this on the Azure OpenAI side where I have the LangChain4J OpenAI and the LangChain4J Azure OpenAI stats, which is exactly the same situation - nearly everyone uses the official SDK.
> The Java version of the Azure AI Inference API had its first release yesterday.

I agree, we should use the official SDK.

jdubois committed:
Initial support (not fully working) for Streaming Chat Model

jdubois committed:
All tests passing for streaming and non-streaming chat models

jdubois committed:
Merge branch 'main' into fix-1719

jdubois committed:
Fix merge issue

jdubois committed:
Remove unnecessary library

jdubois committed:
Add support for embeddings

jdubois commented:
I've added support for all the current models:
- ChatModel
- StreamingChatModel
- EmbeddingModel

I also added tests for everything.

There are a couple of things which aren't perfect:
- Function calling only works with GPT-4. According to https://thenewstack.io/a-comprehensive-guide-to-function-calling-in-llms/ , Mistral and Cohere should support it too, but it doesn't work. I'm guessing this is because they use a different API, and that the Azure AI Inference SDK doesn't support it yet.
- TokenUsage for StreamingChatModel is always 0. As it works for ChatModel, I believe this is currently an API limitation.

To run the tests, you need 2 environment variables:
- `GITHUB_TOKEN` : your GitHub authentication token
- `CI_DELAY_SECONDS_GITHUB_MODELS`: a delay in order not to have rate limitations. I personally put it at 5 seconds.

All tests should be green, but I'll need to do another test tomorrow, as currently I've exhausted my quotas!!

jdubois commented on langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsChatModel.java:
```diff
@@ -0,0 +1,424 @@
+ * <p>
+ * The list of models, as well as the documentation and a playground to test them, can be found at https://github.com/marketplace/models
+ */
+public class GitHubModelsChatModel implements ChatLanguageModel {
```
I also don't like the `GitHubModels` name much, but I'm trying to follow the conventions, and who knows if GitHub won't provide another service one day (I have absolutely no hints on this, but it's so large they could have something else at some point). So I'd keep it like this.

jdubois replied:
Closing as you said you had no strong preference

jdubois commented on langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsChatModel.java:
```diff
@@ -0,0 +1,424 @@
+                                 ChatCompletionsResponseFormat responseFormat,
+                                 List<ChatModelListener> listeners) {
+
+        this.modelName = getOrDefault(modelName, DEFAULT_CHAT_MODEL_NAME);
```
Oh I didn't know that! Let me commit the suggestion and then fix it elsewhere

jdubois committed:
Update langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsChatModel.java

Co-authored-by: LangChain4j <langchain4j@gmail.com>

jdubois committed:
Update langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsChatModel.java

Co-authored-by: LangChain4j <langchain4j@gmail.com>

jdubois committed:
Update langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsChatModel.java

Co-authored-by: LangChain4j <langchain4j@gmail.com>

jdubois committed:
Update langchain4j-github-models/src/main/java/dev/langchain4j/model/github/InternalGitHubModelHelper.java

Co-authored-by: LangChain4j <langchain4j@gmail.com>

jdubois committed:
Merge branch 'fix-1719' of https://github.com/jdubois/langchain4j into fix-1719

jdubois committed:
Merge branch 'main' into fix-1719

jdubois committed:
Merge branch 'fix-1719' of https://github.com/jdubois/langchain4j into fix-1719

jdubois committed:
Fix compilation issues following review

jdubois committed:
Remove unused method

jdubois commented on langchain4j-github-models/src/main/java/dev/langchain4j/model/github/InternalGitHubModelHelper.java:
```diff
@@ -0,0 +1,365 @@
+        }
+    }
+
+    private static String nameFrom(ChatMessage message) {
```
removed

jdubois commented on langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsChatModel.java:
```diff
@@ -0,0 +1,424 @@
+            logger.info("Error generating response, {}", httpResponseException.getValue());
+            FinishReason exceptionFinishReason = contentFilterManagement(httpResponseException, "content_filter");
+            Response<AiMessage> response = Response.from(
+                    aiMessage(httpResponseException.getMessage()),
```
As this is copied on Azure OpenAI, see https://github.com/langchain4j/langchain4j/blob/29299124489179aca7391209ec3b8ae0b9cce28e/langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiChatModel.java#L317 -> my proposal is that we keep it like this at the moment, and then we will discuss and refactor both at the same time, WDYT?

I personally like to have it that field, so people have a warning of what happened (if that's sexual or violence, it's fine to tell them), then I see several people getting annoyed by this, so yes let's refactor this with an exception instead.

jdubois replied:
I see your thumb up, so I created a ticket for this at #1822 -> I'm closing this here, and let's discuss in the ticket.

jdubois commented on langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsStreamingChatModel.java:
```diff
@@ -0,0 +1,566 @@
+ * <p>
+ * The list of models, as well as the documentation and a playground to test them, can be found at https://github.com/marketplace/models
+ */
+public class GitHubModelsStreamingChatModel implements StreamingChatLanguageModel {
```
same as above

jdubois commented:
> @jdubois BTW I did not find `EmbeddingModel`, did you forget to push it?

Not sure I understand you, there is a `GitHubModelsEmbeddingModel`, is it what you are looking for?

jdubois committed:
Fix imports

jdubois commented on langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsStreamingChatModel.java:
```diff
@@ -0,0 +1,566 @@
+        if (useAsyncClient) {
+            this.asyncClient = setupAsyncClient(endpoint, serviceVersion, gitHubToken, timeout, maxRetries, proxyOptions, logRequestsAndResponses, userAgentSuffix, customHeaders);
+        } else {
+            this.client = setupSyncClient(endpoint, serviceVersion, gitHubToken, timeout, maxRetries, proxyOptions, logRequestsAndResponses, userAgentSuffix, customHeaders);
```
Indeed, it doesn't make much sense -> I copied it from Azure OpenAI, and now I realize it was deprecated there!! So indeed, let's remove this. I'll resolve the conversation when it's pushed.

jdubois committed:
Remove the sync client for streaming

jdubois committed:
Fix image details

jdubois commented on langchain4j-github-models/src/main/java/dev/langchain4j/model/github/InternalGitHubModelHelper.java:
```diff
@@ -0,0 +1,365 @@
+                                    throw new IllegalArgumentException("Image URL is not present. Base64 encoded images are not supported at the moment.");
+                                }
+                                ChatMessageImageUrl imageUrl = new ChatMessageImageUrl(imageContent.image().url().toString());
+                                return new ChatMessageImageContentItem(imageUrl);
```
Fixed

jdubois commented on langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsStreamingResponseBuilder.java:
```diff
@@ -0,0 +1,146 @@
+
+    public Response<AiMessage> build() {
+
+        String content = contentBuilder.toString();
```
I just reimplemented it following this algorithm. It makes sense and it's very similar -> I'll close this when this is pushed.
Please note I copied it on the Azure OpenAI module, so it needs to be audited there too.

jdubois commented on langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsStreamingResponseBuilder.java:
```diff
@@ -0,0 +1,146 @@
+        }
+
+        String toolName = toolNameBuilder.toString();
+        if (!toolName.isEmpty()) {
```
As this was modified per the comment above, this should be OK

jdubois committed:
Refactor the algorithm as the LLM can produce both text content and tool calls

jdubois commented on langchain4j-github-models/src/main/java/dev/langchain4j/model/github/InternalGitHubModelHelper.java:
```diff
@@ -0,0 +1,365 @@
+
+    }
+
+    public static ModelServiceVersion getModelServiceVersion(String serviceVersion) {
```
The idea (and this is copied from Azure OpenAI), is that if the user writes a version higher than the actual one, it will fallback to the latest model available. Then, when there's a library upgrade, they will get the new version without refactoring.
I'm not expecting people to use this a lot, as users have no control over the versions deployed on GitHub Models.

jdubois replied:
OK, I have changed it to use ModelServiceVersion directly

jdubois committed:
Use ModelServiceVersion instead of a string

jdubois committed:
Update langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsChatModel.java

Co-authored-by: LangChain4j <langchain4j@gmail.com>

jdubois committed:
Update langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsStreamingChatModel.java

Co-authored-by: LangChain4j <langchain4j@gmail.com>

jdubois committed:
Update langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsEmbeddingModel.java

Co-authored-by: LangChain4j <langchain4j@gmail.com>

jdubois committed:
Update langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsStreamingChatModel.java

Co-authored-by: LangChain4j <langchain4j@gmail.com>

jdubois committed:
Update langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsStreamingChatModel.java

Co-authored-by: LangChain4j <langchain4j@gmail.com>

jdubois committed:
Update langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsStreamingChatModel.java

Co-authored-by: LangChain4j <langchain4j@gmail.com>

jdubois committed:
Update langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsChatModel.java

Co-authored-by: LangChain4j <langchain4j@gmail.com>

jdubois committed:
Update langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsEmbeddingModel.java

Co-authored-by: LangChain4j <langchain4j@gmail.com>

jdubois committed:
Update langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsStreamingChatModel.java

Co-authored-by: LangChain4j <langchain4j@gmail.com>

jdubois committed:
Fix compilation errors

jdubois committed:
Implement knownDimension()

jdubois commented on langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsEmbeddingModel.java:
```diff
@@ -0,0 +1,250 @@
+ * <p>
+ * The list of models, as well as the documentation and a playground to test them, can be found at https://github.com/marketplace/models
+ */
+public class GitHubModelsEmbeddingModel extends DimensionAwareEmbeddingModel {
```
Done

jdubois committed:
Simplify GitHubModelsEmbeddingModelName

jdubois commented on langchain4j-github-models/src/main/java/dev/langchain4j/model/github/GitHubModelsChatModelName.java:
```diff
@@ -0,0 +1,58 @@
+        return modelName;
+    }
+
+    public String modelType() {
```
File has been simplified

jdubois committed:
Optimize imports

jdubois commented:
> BTW what is the plan for `azure-ai-inference` vs `azure-ai-openai`? I read in GItHub Models documentation that users are supposed to still use `azure-ai-inference` when they are ready for production, but we have `langchain4j-azure-open-ai` which uses `azure-ai-openai` and is probably supposed to be used for production (as well)?

It's 2 different products: Azure OpenAI is on Azure, for the big enterprises (you have your own instances, with specific security stuff like Service Principals), while GitHub Models is easy to use for developers (and probably some enterprises too, as it's a lot easier, probably cheaper, and it's the same thing in the end).
The two SDK are pretty similar, then there are some specific things for Azure OpenAI, including security. While GitHub Models is more generic. Those SDKs are mostly generated, so it's not really a duplication effort on the Azure side.

jdubois committed:
Merge branch 'main' into fix-1719

jdubois committed:
Fix dependencies conflicts

jdubois commented:
> Would be also great to have a dedicated documentation page and list Github Models in this table: https://docs.langchain4j.dev/integrations/language-models/
> 
> While documentation can wait a bit, I guess it would be nice to have at least a single example in https://github.com/langchain4j/langchain4j-examples repo for the release (which is happening this Wednesday in the morning)

Here's a PR for the samples: https://github.com/langchain4j/langchain4j-examples/pull/110

jdubois committed:
Add documentation for GitHub Models

jdubois commented:
And I have just pushed the documentation, which links to the examples 👍

jdubois commented:
@langchain4j code, examples and documentation should be OK, I'm crossing my fingers we can have this for tomorrow!

dliubarskyi committed:
Update docs/docs/integrations/language-models/github-models.md

dliubarskyi committed:
Update docs/docs/integrations/language-models/github-models.md

dliubarskyi committed:
Update docs/docs/integrations/language-models/github-models.md
